app/src/main/java/org/wikipedia/descriptions/DescriptionEditFragment.kt

```diff
@@ -0,0 +1,438 @@
+        super.onCreate(savedInstanceState)
+        pageTitle = requireArguments().getParcelable(ARG_TITLE)!!
+        val type = if (pageTitle.description == null) DescriptionEditFunnel.Type.NEW else DescriptionEditFunnel.Type.EXISTING
+        highlightText = requireArguments().getString(ARG_HIGHLIGHT_TEXT)
+        action = requireArguments().getSerializable(ARG_ACTION) as DescriptionEditActivity.Action
+        invokeSource = requireArguments().getSerializable(Constants.INTENT_EXTRA_INVOKE_SOURCE) as InvokeSource
+        requireArguments().getString(ARG_SOURCE_SUMMARY)?.let {
+            sourceSummary = GsonUnmarshaller.unmarshal(PageSummaryForEdit::class.java, it)
+        }
+        requireArguments().getString(ARG_TARGET_SUMMARY)?.let {
+            targetSummary = GsonUnmarshaller.unmarshal(PageSummaryForEdit::class.java, it)
+        }
+        funnel = DescriptionEditFunnel(WikipediaApp.getInstance(), pageTitle, type, invokeSource)
+        funnel.logStart()
+    }
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
+        super.onCreateView(inflater, container, savedInstanceState)
+        _binding = FragmentDescriptionEditBinding.inflate(LayoutInflater.from(context), container, false)
+        loadPageSummaryIfNeeded(savedInstanceState)
+        funnel.logReady()
+        return binding.root
+    }
+
+    override fun onDestroyView() {
+        binding.fragmentDescriptionEditView.setCallback(null)
+        _binding = null
+        super.onDestroyView()
+    }
+
+    override fun onDestroy() {
+        cancelCalls()
+        super.onDestroy()
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putString(ARG_DESCRIPTION, binding.fragmentDescriptionEditView.description)
+        outState.putBoolean(ARG_REVIEWING, binding.fragmentDescriptionEditView.showingReviewContent())
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        if (requestCode == Constants.ACTIVITY_REQUEST_DESCRIPTION_EDIT_SUCCESS && activity != null) {
+            callback()?.onDescriptionEditSuccess()
+        } else if (requestCode == Constants.ACTIVITY_REQUEST_VOICE_SEARCH &&
+                resultCode == Activity.RESULT_OK && data != null &&
+                data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS) != null) {
+            val text = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)!![0]
+            binding.fragmentDescriptionEditView.setDescription(text)
+        }
+    }
+
+    private fun cancelCalls() {
+        csrfClient?.cancel()
+        csrfClient = null
+        disposables.clear()
+    }
+
+    private fun loadPageSummaryIfNeeded(savedInstanceState: Bundle?) {
+        binding.fragmentDescriptionEditView.showProgressBar(true)
+        if (invokeSource == InvokeSource.PAGE_ACTIVITY && sourceSummary!!.extractHtml!!.isEmpty()) {
```

Comment:
`sourceSummary!!.extractHtml!!.isEmpty()` can be `sourceSummary?.extractHtml.isNullOrEmpty()`